src/Compiler/Symbols/FSharpDiagnostic.fs

```diff
@@ -9,19 +9,104 @@ namespace FSharp.Compiler.Diagnostics
 
 open System
 
-open Internal.Utilities.Library  
+open FSharp.Compiler.CheckExpressions
+open FSharp.Compiler.ConstraintSolver
+open FSharp.Compiler.SignatureConformance
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreeBasics
+open FSharp.Compiler.TypedTreeOps
+open Internal.Utilities.Library
 open Internal.Utilities.Library.Extras
 
 open FSharp.Core.Printf
-open FSharp.Compiler 
+open FSharp.Compiler
 open FSharp.Compiler.CompilerDiagnostics
 open FSharp.Compiler.Diagnostics
 open FSharp.Compiler.DiagnosticsLogger
 open FSharp.Compiler.Text
 open FSharp.Compiler.Text.Position
 open FSharp.Compiler.Text.Range
 
-type FSharpDiagnostic(m: range, severity: FSharpDiagnosticSeverity, message: string, subcategory: string, errorNum: int, numberPrefix: string) =
+[<RequireQualifiedAccess; Experimental("This FCS API is experimental and subject to change.")>]
+type DiagnosticContextInfo =
+    | NoContext
+    | IfExpression
+    | OmittedElseBranch
+    | ElseBranchResult
+    | RecordFields
+    | TupleInRecordFields
+    | CollectionElement
+    | ReturnInComputationExpression
+    | YieldInComputationExpression
+    | RuntimeTypeTest
+    | DowncastUsedInsteadOfUpcast
+    | FollowingPatternMatchClause
+    | PatternMatchGuard
+    | SequenceExpression
```

Comment:
Are any of these diagnostics emitted as warning by default? (i.e. can be emitted for multiple ranges in file/project)
If yes, can we test how much more do we allocate in tooling scenarios for those?

Reply:
@vzarytovskii, 

`DiagnosticContextInfo` is just a context in which a diagnostic with a type error occurred.

Copied from internal api:
https://github.com/dotnet/fsharp/blob/d95e8f3e0c6341780718a30ca86acae3d28a48a6/src/Compiler/Checking/ConstraintSolver.fsi#L73-L117

Usage example from prototype: 
https://github.com/JetBrains/resharper-fsharp/blob/fd9dd3c7342de371b7a0af6a77b88ed00109ad90/ReSharper.FSharp/src/FSharp.Psi.Daemon/src/Stages/FcsErrorsStageProcessBase.fs#L183-L195

Reply:
`ExtendedData` is created here
https://github.com/dotnet/fsharp/blob/9cf7dcb881e7c3c482d6abcaa6624f3fe672f5f1/src/Compiler/Symbols/FSharpDiagnostic.fs#L164-L191

Reply:
Discussed offline